File: 021_merge_two_sorted_linked_lists/solution.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    def __str__(self):
        return str(self.val)

    def __repr__(self):
        return str(self.val)

    @property
    def length(self):
        res = []
        curr = self
        while curr:
            res.append(curr)
            curr = curr.next
        return res


def print_list(root: ListNode):
    curr = root
    while curr:
        print(curr)
        curr = curr.next


list1 = [1, 2, 4]
list2 = [1, 3, 4]

# lst 1:
root1 = ListNode(val=1)
root1.next = ListNode(val=3)
root1.next.next = ListNode(val=4)

# lst 2:
root2 = ListNode(val=2)
root2.next = ListNode(val=5)
root2.next.next = ListNode(val=6)


def merge(root1: ListNode, root2: ListNode):
    res_root = ListNode(val=None)
    curr = res_root
    curr1 = root1
    curr2 = root2

    while curr1 and curr2:
        if curr1.val > curr2.val:
            curr.next = curr2
            curr2 = curr2.next
        else:
            curr.next = curr1
            curr1 = curr1.next
        curr = curr.next

    if curr1:
        curr.next = curr1
    if curr2:
        curr.next = curr2

    return res_root.next


print_list(merge(root1, root2))

File: 021_merge_two_sorted_linked_lists/solution.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	result := &ListNode{}
	curr := result
	curr1 := list1
	curr2 := list2

	for curr1 != nil && curr2 != nil {
		if curr1.Val > curr2.Val {
			curr.Next = curr2
			curr2 = curr2.Next
		} else {
			curr.Next = curr1
			curr1 = curr1.Next
		}
		curr = curr.Next
	}
	if curr1 != nil {
		curr.Next = curr1
	}
	if curr2 != nil {
		curr.Next = curr2
	}
	return result.Next
}

func main() {
	node1 := ListNode{Val: 1, Next: &ListNode{Val: 2, Next: &ListNode{Val: 4}}}
	node2 := ListNode{Val: 1, Next: &ListNode{Val: 3, Next: &ListNode{Val: 4}}}

	fmt.Println(mergeTwoLists(&node1, &node2))
}
